core/src/mindustry/mod/ContentParser.java

```diff
@@ -512,13 +513,30 @@ private <T> T internalRead(Class<T> type, Class elementType, JsonValue jsonData,
                 }
 
                 if(value.has("controller") || value.has("aiController")){
-                    unit.aiController = supply(resolve(value.getString("controller", value.getString("aiController", "")), FlyingAI.class));
+                    JsonValue con = value.has("controller") ? value.get("controller") : value.get("aicontroller");
+                    if(con.isObject()){
+                        AIController ai = parser.readValue(AIController.class, con);
+                        unit.aiController = () -> ai;
+                    }else{
+                        unit.aiController = supply(resolve(con.isString() ? con.asString() : "", FlyingAI.class));
+                    }
                     value.remove("controller");
                 }
 
                 if(value.has("defaultController")){
-                    var sup = supply(resolve(value.getString("defaultController"), FlyingAI.class));
-                    unit.controller = u -> sup.get();
+                    JsonValue con = value.get("defaultController");
+                    if(con.isObject()){
+                        var sup = supply(resolve(con.getString("type"), FlyingAI.class));
+                        con.remove("type");
+                        unit.controller = u -> {
+                            AIController ai = sup.get();
+                            readFields(ai, con); //TODO is there a way to make it only need to read fields once?
```

Comment:
Is there a better way to do this? Reading fields every time the unit is spawned is probably very inefficient.
I want to get the code to a cleaner state before I move on to testing it.

Reply:
Possibly making AI types implement Cloneable, reading in one instance, and making it return a clone()-ed instance?